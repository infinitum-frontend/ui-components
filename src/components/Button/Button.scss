.inf-button {
  font-family: var(--inf-font-family-base);
  font-size: 14px;
  font-weight: 400;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: -webkit-max-content;
  width: max-content;
  padding: 0 16px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-decoration: none;
  border: 1px solid transparent;
  box-shadow: 0 0 0 2px transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: var(--inf-border-radius-medium);
  transition: 0.2s;
  background: none;
  position: relative;

  &:hover {
    border-color: transparent;
  }

  &:focus-visible {
    // outline: none;
  }

  &:focus {
    // TODO: border: 2px solid #7abfff;
  }

  &--variant-primary {
    background-color: var(--inf-color-primary);
    color: #fff;

    &:hover {
      background-color: var(--inf-color-primary-light);
    }

    &:active {
      background-color: var(--inf-color-primary-dark);
    }
  }

  &--variant-secondary {
    background-color: #738f99;
    color: #fff;

    &:hover {
      background-color: #b8d1d9;
    }

    &:active {
      background-color: #46626b;
    }
  }

  &--variant-tertiary {
    color: #46626b;
    border: 1px solid #46626b;

    &:hover {
      background-color: #e6f1f5;
    }

    &:active {
      background-color: #738f99;
    }
  }

  &--variant-ghost {
    color: var(--inf-color-text-secondary);

    &:hover {
      color: var(--inf-color-text-tertiary);
    }

    &:active {
      color: var(--inf-color-text-tertiary);
    }
  }

  &:disabled {
    background-color: var(--inf-color-background-control-disabled);
    border-color: var(--inf-color-background-control-disabled);
    color: var(--inf-color-text-quaternary);
    pointer-events: none;
    cursor: default;
  }

  &--block {
    width: 100%;
  }

  &--loading {
    pointer-events: none;
    cursor: default;

    .inf-button__content {
      visibility: hidden;
    }
  }

  &--square {
    padding: 0;
  }

  &--size-small {
    height: var(--inf-control-height-small);

    &.inf-button--square {
      width: var(--inf-control-height-small);
    }
  }

  &--size-medium {
    height: var(--inf-control-height-medium);

    &.inf-button--square {
      width: var(--inf-control-height-medium);
    }
  }

  &--size-large {
    height: var(--inf-control-height-large);

    &.inf-button--square {
      width: var(--inf-control-height-large);
    }
  }

  &__loader {
    position: absolute;
  }

  &__content {
    flex-shrink: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    display: flex;
    align-items: center;
  }

  &__before,
  &__after,
  &__icon {
    display: flex;
  }

  .inf-button__after:not(:first-child),
  .inf-button__before:not(:last-child) + .inf-button__text {
    margin-left: var(--inf-space-xsmall);
  }
}
