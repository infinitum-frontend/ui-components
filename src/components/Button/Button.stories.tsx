// eslint-disable-next-line @typescript-eslint/no-unused-vars
import * as React from 'react'
import { Button } from './index'
import { Space } from '../Space'
import { StoryObj, Meta, StoryFn } from '@storybook/react'
import { ReactComponent as ArrowDownIcon } from 'Icons/chevron-down.svg'
import { ReactComponent as DownloadIcon } from 'Icons/download.svg'

const ComponentMeta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  args: {
    children: 'Кнопка',
    as: 'button'
  },
  parameters: {
    docs: {
      source: {
        excludeDecorators: true
      }
    }
  }
}

export default ComponentMeta

const Template: StoryFn<typeof Button> = ({ ...args }) => <Button {...args} />

export const Playground = {
  render: Template
}

export const Secondary = {
  render: Template,

  args: {
    variant: 'secondary'
  }
}

export const Ghost = {
  render: Template,

  args: {
    variant: 'ghost'
  }
}

export const GhostWithIcon = {
  render: Template,

  args: {
    variant: 'ghost',
    after: <DownloadIcon />
  }
}

export const Loading = {
  render: Template,

  args: {
    loading: true
  }
}

export const Disabled = {
  render: Template,

  args: {
    disabled: true
  }
}

export const Icon: StoryObj<typeof Button> = {
  render: (args) => (
    <>
      <Button before={<ArrowDownIcon />} {...args}>
        Icon Before
      </Button>
      <Button after={<ArrowDownIcon />} {...args}>
        Icon After
      </Button>
      <Button before={<ArrowDownIcon />} after={<ArrowDownIcon />} {...args}>
        Icon Before and After
      </Button>
      <Button icon={<ArrowDownIcon />} {...args} />
    </>
  ),

  decorators: [
    (Story) => (
      <Space gap="small" direction="horizontal">
        {Story()}
      </Space>
    )
  ]
}

export const Sizes: StoryObj<typeof Button> = {
  render: (args) => (
    <>
      <Button {...args} size="large" />
      <Button {...args} size="medium" />
      <Button {...args} size="small" />
    </>
  ),

  decorators: [
    (Story) => (
      <Space gap="small" direction="horizontal">
        {Story()}
      </Space>
    )
  ]
}

export const LinkButton: StoryObj<typeof Button> = {
  render: (args) => (
    <Button href="https://specdep.ru/" target="_blank" {...args} />
  ),

  args: {
    as: 'a',
    children: 'Кнопка-ссылка'
  }
}
