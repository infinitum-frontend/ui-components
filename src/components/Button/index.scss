.inf-button {
  font-family: Helvetica, Arial, sans-serif;
  font-weight: 300;
  font-size: 18px;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  width: -webkit-max-content;
  width: max-content;
  padding: 0 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-decoration: none;
  border: 1px solid transparent;
  box-shadow: 0 0 0 2px transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-radius: 50px; // button-border-radius
  transition: .2s;

  &:hover {
    filter: brightness(110%);
  }

  &:active {
    filter: brightness(90%);
  }

  &:focus-visible {
    outline: none;
  }

  &--variant-primary {
    background-color: var(--inf-color-primary);
    color: #fff;
  }

  &--variant-secondary {
    background: #f6f7f8;
    color: #646a89;
  }

  &--size-small {
    height: var(--inf-control-height-small);
    font-size: 14px;
  }

  &--size-medium {
    height: var(--inf-control-height-medium);
  }

  &--size-large {
    height: var(--inf-control-height-large);
  }
}
