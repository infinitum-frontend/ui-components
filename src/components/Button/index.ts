export { default as InfButton } from './InfButton'
