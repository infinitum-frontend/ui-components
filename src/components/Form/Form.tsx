// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  ComponentPropsWithRef,
  CSSProperties,
  FormEventHandler,
  forwardRef,
  ReactElement,
  useRef
} from 'react'
import cn from 'classnames'
import { Space, SpaceProps } from 'Components/Space'
import FormContext from './context/form'
import FormLabel from './components/FormLabel'
import FormGroup from './components/FormGroup'
import FormHint from './components/FormHint'
import FormItem from './components/FormItem'
import FormErrorMessage from './components/FormErrorMessage'
import FormAction from './components/FormAction'
import { mergeRefs } from 'react-merge-refs'

export interface FormProps extends ComponentPropsWithRef<'form'> {
  /** Расстояние между блоками */
  gap?: SpaceProps['gap']
  /** Состояние недоступности */
  disabled?: boolean
  /** Ширина лейбла (валидное значение css-width) */
  labelWidth?: CSSProperties['width']
}

const Form = forwardRef<HTMLFormElement, FormProps>(
  (
    {
      gap = 'medium',
      children,
      labelWidth = '',
      onSubmit,
      className,
      disabled,
      ...props
    },
    ref
  ): ReactElement => {
    const innerRef = useRef<HTMLFormElement>(null)

    const handleSubmit: FormEventHandler<HTMLFormElement> = (e) => {
      e.preventDefault()

      const form = e.target as HTMLFormElement
      const isValid = form.checkValidity()
      if (!isValid) {
        return false
      }

      onSubmit?.(e)
    }

    const context = {
      labelWidth,
      form: innerRef,
      disabled
    }

    return (
      <FormContext.Provider value={context}>
        <Space
          as={'form'}
          gap={gap}
          onSubmit={handleSubmit}
          noValidate={true}
          ref={mergeRefs([innerRef, ref])}
          className={cn('inf-form', className, {})}
          {...props}
        >
          {children}
        </Space>
      </FormContext.Provider>
    )
  }
)

Form.displayName = 'Form'

/** Компонент для создания интерактивных и удобных форм ввода, упрощающий пользователям ввод и отправку данных с проверками и настраиваемыми полями */
export default Object.assign(Form, {
  Group: FormGroup,
  Item: FormItem,
  Label: FormLabel,
  Hint: FormHint,
  ErrorMessage: FormErrorMessage,
  Action: FormAction
})
