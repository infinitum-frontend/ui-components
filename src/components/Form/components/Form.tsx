// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  ComponentPropsWithRef,
  CSSProperties,
  FormEventHandler,
  forwardRef,
  ReactElement
} from 'react'
import cn from 'classnames'
import FormLabel from 'Components/Form/components/Label'
import FormGroup from 'Components/Form/components/Group'
import FormHint from 'Components/Form/components/Hint'
import Space, { SpaceProps } from 'Components/Space/Space'
import FormContext from 'Components/Form/context/form'
import FormItem from 'Components/Form/components/Item'
import FormErrorMessage from 'Components/Form/components/ErrorMessage'

export interface FormProps extends ComponentPropsWithRef<'form'> {
  /**
   * Расстояние между блоками
   */
  gap?: SpaceProps['gap']
  /** Ширина лейбла (валидное значение css-width) */
  labelWidth?: CSSProperties['width']
}

const Form = forwardRef<HTMLFormElement, FormProps>(
  (
    {
      gap = 'medium',
      children,
      labelWidth = '',
      onSubmit,
      className,
      ...props
    },
    ref
  ): ReactElement => {
    const handleSubmit: FormEventHandler<HTMLFormElement> = (e) => {
      e.preventDefault()
      onSubmit?.(e)
    }

    const context = {
      labelWidth
    }

    return (
      <FormContext.Provider value={context}>
        <Space
          as={'form'}
          gap={gap}
          onSubmit={handleSubmit}
          ref={ref}
          className={cn('inf-form', className, {})}
          {...props}
        >
          {children}
        </Space>
      </FormContext.Provider>
    )
  }
)

Form.displayName = 'Form'

/** Компонент для создания интерактивных и удобных форм ввода, упрощающий пользователям ввод и отправку данных с проверками и настраиваемыми полями */
export default Object.assign(Form, {
  Group: FormGroup,
  Item: FormItem,
  Label: FormLabel,
  Hint: FormHint,
  ErrorMessage: FormErrorMessage
})
