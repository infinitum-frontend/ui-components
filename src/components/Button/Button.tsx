// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { ElementType, forwardRef, ReactElement, ReactNode } from 'react'
import cn from 'classnames'
import { Loader } from '../Loader'
import './Button.scss'
import { PolymorphicComponent, PolymorphicRef } from '~/src/utils/types'

export interface ButtonProps {
  /**
   * Содержимое
   */
  children?: ReactNode
  /**
   * Дополнительный className
   */
  className?: string
  /**
   * Вариант оформления
   */
  variant?: 'primary' | 'secondary' | 'tertiary' | 'ghost'
  // outlined?: boolean TODO:
  // plain?: boolean TODO:
  /**
   * Размер
   */
  size?: 'small' | 'medium' | 'large'
  /**
   * Состояние недоступности
   */
  disabled?: boolean
  /**
   * Состояние загрузки
   */
  loading?: boolean
  /**
   * Занимает всю ширину контейнера
   */
  block?: boolean
  /**
   * Иконка (без текста)
   */
  icon?: ReactNode
  /**
   * Контент слева от текста
   */
  before?: ReactNode
  /**
   * Контент справа от текста
   */
  after?: ReactNode
}

function BaseButton<C extends ElementType = 'button'>(
  props: PolymorphicComponent<C, ButtonProps>,
  ref: PolymorphicRef<C>
): ReactElement {
  const {
    children,
    className,
    as = 'button',
    variant = 'primary',
    size = 'medium',
    loading = false,
    block = false,
    before,
    after,
    icon,
    ...rest
  } = props

  const Component = as

  return (
    <Component
      ref={ref}
      className={cn(
        'inf-button',
        className,
        `inf-button--variant-${variant as string}`,
        `inf-button--size-${size as string}`,
        {
          'inf-button--block': block,
          'inf-button--loading': loading,
          'inf-button--square': icon
        }
      )}
      type="button"
      {...rest}
    >
      <span className="inf-button__content">
        {icon ? (
          <span className="inf-button__icon">{icon}</span>
        ) : (
          <>
            {before && <span className="inf-button__before">{before}</span>}
            <span className="inf-button__text">{children}</span>
            {after && <span className="inf-button__after">{after}</span>}
          </>
        )}
      </span>

      {loading && (
        <Loader className="inf-button__loader" size="compact" variant="unset" />
      )}
    </Component>
  )
}

const Button = forwardRef(BaseButton)

/** Компонент кнопки, используемый для инициализации различных действий */
export default Button as typeof BaseButton
