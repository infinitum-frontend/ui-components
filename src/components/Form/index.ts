import Form, { FormProps } from './Form'
import { FormGroupProps } from './components/FormGroup'

export { Form }
export type { FormProps, FormGroupProps }
