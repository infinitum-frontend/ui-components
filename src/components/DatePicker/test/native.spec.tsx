import { describe, it, expect, vi } from 'vitest'
import { renderComponent } from '../../../../testSetup'
import { act, screen } from '@testing-library/react'
import { DatePicker, NativeDatePicker, DatePickerInline } from '../index'
import userEvent from '@testing-library/user-event'
import { Form } from '../../Form'
import { Button } from '../../Button'
const user = userEvent.setup()

describe('NativeDatePicker', () => {
  it('should match snapshop', () => {
    renderComponent(<NativeDatePicker />)
    expect(screen.queryByRole('calendar')).toMatchSnapshot()
  })

  it('should support type', () => {
    renderComponent(<NativeDatePicker type={'week'} />)
    expect(screen.queryByRole('calendar')).toHaveAttribute('type', 'week')
  })

  it('should support onChange', async () => {
    const onChange = vi.fn()
    renderComponent(<NativeDatePicker onChange={onChange} />)

    act(() => (screen.queryByRole('calendar') as HTMLInputElement).focus())
    await user.keyboard('1984-09-14')

    expect(onChange).toHaveBeenCalledOnce()
  })

  it('should support value', () => {
    renderComponent(<NativeDatePicker value={'1984-09-14'} />)
    expect(screen.queryByRole('calendar')).toHaveValue('1984-09-14')
  })

  it('should support className', () => {
    renderComponent(<NativeDatePicker className={'custom-class'} />)
    expect(screen.queryByRole('calendar')).toHaveClass(
      'custom-class inf-native-datepicker'
    )
  })

  it('should support required and id', () => {
    renderComponent(<NativeDatePicker required={true} id={'calendar'} />)
    expect(screen.queryByRole('calendar')).toBeRequired()
    expect(screen.queryByRole('calendar')).toHaveAttribute('id', 'calendar')
  })
})

describe('Datepicker in Form', () => {
  it('should support attributes', async () => {
    const errorMessage = 'Ошибка валидации'
    renderComponent(
      <Form>
        <Form.Group required customValidationMessage={errorMessage}>
          <NativeDatePicker />
        </Form.Group>
        <Button type={'submit'} />
      </Form>
    )

    const datepicker = screen.queryByRole('calendar') as HTMLTextAreaElement

    expect(datepicker).toHaveAttribute('id')
    expect(datepicker).toHaveAttribute('required')
    expect(datepicker).toHaveAttribute('aria-required', 'true')
    expect(datepicker).toHaveAttribute('aria-invalid', 'false')

    await user.click(screen.queryByRole('button') as HTMLButtonElement)

    expect(datepicker).toHaveAttribute('aria-invalid', 'true')
    expect(screen.queryByText(errorMessage)).toBeInTheDocument()

    act(() => {
      datepicker.focus()
    })
    await user.keyboard('1984-09-14')

    expect(datepicker).toHaveAttribute('aria-invalid', 'false')
    expect(datepicker.validationMessage).toBe('')
  })
})

describe('should support withTodayButton prop', () => {
  it('should render today button if withTodayButton prop is passed', async () => {
    renderComponent(<DatePicker withTodayButton={true} />)

    const user = userEvent.setup()
    const input = screen.queryByRole('textbox') as HTMLInputElement
    await user.click(input)

    expect(screen.queryByText('Сегодня')).toBeInTheDocument()
  })

  it('should render today button if withTodayButton prop is passed to InlineDatePicker', () => {
    renderComponent(<DatePickerInline withTodayButton={true} />)
    expect(screen.queryByText('Сегодня')).toBeInTheDocument()
  })
})
