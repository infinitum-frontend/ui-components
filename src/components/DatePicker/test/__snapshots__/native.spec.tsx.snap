// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NativeDatePicker > should match snapshop 1`] = `
<input
  autocomplete="off"
  class="inf-native-datepicker inf-text-field inf-text-field--br-regular inf-text-field--size-medium"
  role="calendar"
  type="date"
  value=""
/>
`;
